File: swift-bridge/Sources/AVPlayerBridge/PlayerLooper.swift
import AVFoundation
import Foundation

@_cdecl("av_player_looper_create")
public func av_player_looper_create(
    _ playerPtr: UnsafeMutableRawPointer,
    _ templateItemPtr: UnsafeMutableRawPointer,
    _ useLoopRange: Bool,
    _ startValue: Int64,
    _ startTimescale: Int32,
    _ startKind: Int32,
    _ durationValue: Int64,
    _ durationTimescale: Int32,
    _ durationKind: Int32,
    _ itemOrderingRaw: Int32,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    let player = Unmanaged<AVQueuePlayer>.fromOpaque(playerPtr).takeUnretainedValue()
    let templateItem = Unmanaged<AVPlayerItem>.fromOpaque(templateItemPtr).takeUnretainedValue()
    let loopRange = useLoopRange
        ? CMTimeRange(
            start: cmTime(value: startValue, timescale: startTimescale, kind: startKind),
            duration: cmTime(value: durationValue, timescale: durationTimescale, kind: durationKind)
        )
        : .invalid

    if #available(macOS 14.0, *) {
        let itemOrdering: AVPlayerLooper.ItemOrdering = itemOrderingRaw == 1
            ? .loopingItemsFollowExistingItems
            : .loopingItemsPrecedeExistingItems
        return Unmanaged.passRetained(
            AVPlayerLooper(player: player, templateItem: templateItem, timeRange: loopRange, existingItemsOrdering: itemOrdering)
        ).toOpaque()
    }

    if itemOrderingRaw != 0 {
        outErrorMessage?.pointee = ffiString("existingItemsOrdering requires macOS 14.0+")
        return nil
    }

    return Unmanaged.passRetained(AVPlayerLooper(player: player, templateItem: templateItem, timeRange: loopRange)).toOpaque()
}

@_cdecl("av_player_looper_release")
public func av_player_looper_release(_ looperPtr: UnsafeMutableRawPointer?) {
    guard let looperPtr else { return }
    Unmanaged<AVPlayerLooper>.fromOpaque(looperPtr).release()
}

@_cdecl("av_player_looper_info_json")
public func av_player_looper_info_json(
    _ looperPtr: UnsafeMutableRawPointer,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    let looper = Unmanaged<AVPlayerLooper>.fromOpaque(looperPtr).takeUnretainedValue()
    let payload = PlayerLooperInfoPayload(
        status: Int32(looper.status.rawValue),
        errorMessage: looper.error?.localizedDescription,
        loopCount: looper.loopCount,
        loopingItemCount: looper.loopingPlayerItems.count
    )
    do {
        return ffiString(try avpEncodeJSON(payload))
    } catch {
        outErrorMessage?.pointee = ffiString(error.localizedDescription)
        return nil
    }
}

@_cdecl("av_player_looper_disable_looping")
public func av_player_looper_disable_looping(_ looperPtr: UnsafeMutableRawPointer) {
    let looper = Unmanaged<AVPlayerLooper>.fromOpaque(looperPtr).takeUnretainedValue()
    looper.disableLooping()
}

@_cdecl("av_player_looper_looping_item_count")
public func av_player_looper_looping_item_count(_ looperPtr: UnsafeMutableRawPointer) -> Int32 {
    let looper = Unmanaged<AVPlayerLooper>.fromOpaque(looperPtr).takeUnretainedValue()
    return Int32(looper.loopingPlayerItems.count)
}

@_cdecl("av_player_looper_copy_looping_item_at_index")
public func av_player_looper_copy_looping_item_at_index(
    _ looperPtr: UnsafeMutableRawPointer,
    _ index: Int32
) -> UnsafeMutableRawPointer? {
    let looper = Unmanaged<AVPlayerLooper>.fromOpaque(looperPtr).takeUnretainedValue()
    let items = looper.loopingPlayerItems
    guard index >= 0, Int(index) < items.count else { return nil }
    return Unmanaged.passRetained(items[Int(index)]).toOpaque()
}

File: src/player_looper.rs
#![allow(clippy::missing_errors_doc, clippy::must_use_candidate)]

use core::ffi::{c_char, c_void};
use core::ptr;

use serde::Deserialize;

use crate::error::{from_swift, AVPlayerError};
use crate::ffi;
use crate::player::PlayerItem;
use crate::queue_player::QueuePlayer;
use crate::time::TimeRange;
use crate::util::parse_json_and_free;

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PlayerLooperInfoPayload {
    status: i32,
    error_message: Option<String>,
    loop_count: i64,
    looping_item_count: i64,
}

/// Mirrors the `AVPlayer` framework counterpart for `PlayerLooperStatus`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum PlayerLooperStatus {
    /// Mirrors the `AVPlayer` framework case `Unknown`.
    Unknown,
    /// Mirrors the `AVPlayer` framework case `Ready`.
    Ready,
    /// Mirrors the `AVPlayer` framework case `Failed`.
    Failed,
    /// Mirrors the `AVPlayer` framework case `Cancelled`.
    Cancelled,
}

impl PlayerLooperStatus {
    /// Mirrors the `AVPlayer` framework constant `fn`.
    #[must_use]
    pub const fn from_raw(raw: i32) -> Self {
        match raw {
            1 => Self::Ready,
            2 => Self::Failed,
            3 => Self::Cancelled,
            _ => Self::Unknown,
        }
    }
}

/// Mirrors the `AVPlayer` framework counterpart for `PlayerLooperItemOrdering`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum PlayerLooperItemOrdering {
    /// Mirrors the `AVPlayer` framework case `LoopingItemsPrecedeExistingItems`.
    LoopingItemsPrecedeExistingItems,
    /// Mirrors the `AVPlayer` framework case `LoopingItemsFollowExistingItems`.
    LoopingItemsFollowExistingItems,
}

impl PlayerLooperItemOrdering {
    /// Mirrors the `AVPlayer` framework constant `fn`.
    #[must_use]
    pub const fn as_raw(self) -> i32 {
        match self {
            Self::LoopingItemsPrecedeExistingItems => 0,
            Self::LoopingItemsFollowExistingItems => 1,
        }
    }
}

/// Mirrors the `AVPlayer` framework counterpart for `PlayerLooper`.
#[derive(Debug)]
pub struct PlayerLooper {
    ptr: *mut c_void,
}

impl Drop for PlayerLooper {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::av_player_looper_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

// SAFETY: AVPlayerLooper ObjC handles are safe to transfer across thread
// boundaries; method calls are internally dispatched safely.
unsafe impl Send for PlayerLooper {}

impl PlayerLooper {
    /// Calls the `AVPlayer` framework counterpart for `new`.
    pub fn new(player: &QueuePlayer, template_item: &PlayerItem) -> Result<Self, AVPlayerError> {
        Self::with_time_range_and_ordering(
            player,
            template_item,
            None,
            PlayerLooperItemOrdering::LoopingItemsPrecedeExistingItems,
        )
    }

    /// Calls the `AVPlayer` framework counterpart for `with_time_range`.
    pub fn with_time_range(
        player: &QueuePlayer,
        template_item: &PlayerItem,
        loop_range: TimeRange,
    ) -> Result<Self, AVPlayerError> {
        Self::with_time_range_and_ordering(
            player,
            template_item,
            Some(loop_range),
            PlayerLooperItemOrdering::LoopingItemsPrecedeExistingItems,
        )
    }

    /// Calls the `AVPlayer` framework counterpart for `with_time_range_and_ordering`.
    pub fn with_time_range_and_ordering(
        player: &QueuePlayer,
        template_item: &PlayerItem,
        loop_range: Option<TimeRange>,
        item_ordering: PlayerLooperItemOrdering,
    ) -> Result<Self, AVPlayerError> {
        let (
            start_value,
            start_timescale,
            start_kind,
            duration_value,
            duration_timescale,
            duration_kind,
        ) = loop_range.map_or((0, 0, 1, 0, 0, 1), |loop_range| {
            let (start_value, start_timescale, start_kind) = loop_range.start.to_raw();
            let (duration_value, duration_timescale, duration_kind) = loop_range.duration.to_raw();
            (
                start_value,
                start_timescale,
                start_kind,
                duration_value,
                duration_timescale,
                duration_kind,
            )
        });
        let mut err: *mut c_char = ptr::null_mut();
        let ptr = unsafe {
            ffi::av_player_looper_create(
                player.ptr,
                template_item.ptr,
                loop_range.is_some(),
                start_value,
                start_timescale,
                start_kind,
                duration_value,
                duration_timescale,
                duration_kind,
                item_ordering.as_raw(),
                &mut err,
            )
        };
        if ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::PLAYER_CREATE_FAILED, err) });
        }
        Ok(Self { ptr })
    }

    fn info(&self) -> Result<PlayerLooperInfoPayload, AVPlayerError> {
        let mut err: *mut c_char = ptr::null_mut();
        let json_ptr = unsafe { ffi::av_player_looper_info_json(self.ptr, &mut err) };
        if json_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        parse_json_and_free(json_ptr)
    }

    /// Calls the `AVPlayer` framework counterpart for `status`.
    pub fn status(&self) -> Result<PlayerLooperStatus, AVPlayerError> {
        Ok(PlayerLooperStatus::from_raw(self.info()?.status))
    }

    /// Calls the `AVPlayer` framework counterpart for `error`.
    pub fn error(&self) -> Result<Option<String>, AVPlayerError> {
        Ok(self.info()?.error_message)
    }

    /// Calls the `AVPlayer` framework counterpart for `loop_count`.
    pub fn loop_count(&self) -> Result<i64, AVPlayerError> {
        Ok(self.info()?.loop_count)
    }

    /// Calls the `AVPlayer` framework counterpart for `looping_items`.
    pub fn looping_items(&self) -> Result<Vec<PlayerItem>, AVPlayerError> {
        let count = self.info()?.looping_item_count;
        if count < 0 {
            return Err(AVPlayerError::OperationFailed(format!(
                "looping item count unexpectedly negative: {count}"
            )));
        }
        let count = usize::try_from(count).map_err(|error| {
            AVPlayerError::OperationFailed(format!(
                "looping item count exceeds addressable size: {error}"
            ))
        })?;

        let mut items = Vec::with_capacity(count);
        for index in 0..count {
            let ptr = unsafe {
                ffi::av_player_looper_copy_looping_item_at_index(
                    self.ptr,
                    i32::try_from(index).map_err(|error| {
                        AVPlayerError::OperationFailed(format!(
                            "looping item index exceeds bridge range: {error}"
                        ))
                    })?,
                )
            };
            if ptr.is_null() {
                return Err(AVPlayerError::OperationFailed(format!(
                    "bridge returned null looping item at index {index}"
                )));
            }
            items.push(PlayerItem { ptr });
        }
        Ok(items)
    }

    /// Calls the `AVPlayer` framework counterpart for `disable_looping`.
    pub fn disable_looping(&self) {
        unsafe { ffi::av_player_looper_disable_looping(self.ptr) };
    }
}
